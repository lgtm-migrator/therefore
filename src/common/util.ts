import type { Traversable } from '@skyleague/axioms'

export async function awaitAll<T>(xs: Traversable<Promise<T>> | Traversable<T>): Promise<T[]>
export async function awaitAll<T, U = T>(xs: Traversable<T>, f: (x: T) => Promise<U> | U): Promise<U[]>
export async function awaitAll<T, U = T>(xs: Traversable<T>, f?: (x: T) => Promise<U> | U): Promise<U[]> {
    const results: U[] = []
    if (f !== undefined) {
        for (const x of xs) {
            results.push(await f(x))
        }
    } else {
        for (const x of xs) {
            results.push(await (x as unknown as Promise<U>))
        }
    }
    return results
}
