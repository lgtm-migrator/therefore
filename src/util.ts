export function isDefined<TValue>(value: TValue | null | undefined): value is TValue {
    return value !== null && value !== undefined
}

export function filterUndefined<T, U>(obj: Partial<T> & U): U {
    const lobj: Record<string, unknown> = obj
    for (const key in lobj) {
        if (lobj[key] === undefined) {
            delete lobj[key]
        }
    }
    return obj
}
