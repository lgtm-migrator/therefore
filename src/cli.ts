import { FileDefinition, TypescriptReference } from './definition'
import { toTypescriptDefinition } from './typescript'
import { toJsonSchema } from './schema'
import { renderTemplate } from './template'
import { isDefined } from './util'
import {
    ArrayType,
    DictType,
    IntersectionType,
    isShorthand,
    ObjectType,
    RefType,
    ThereforeTypes,
    TupleType,
    UnionType,
} from './types/composite'
import { schema } from './therefore'
import { EnumType } from './types/enum'

import decamelize from 'decamelize'
import execa from 'execa'
import fastGlob from 'fast-glob'

import path from 'path'
import fs from 'fs'

// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment
const { version, name: packageName } = require('../package.json')
const generatedBy = `Generated by ${packageName as string}`

export function isEnum(obj: ThereforeTypes | { [schema.type]: string }): obj is EnumType {
    return obj[schema.type] === 'enum'
}

export function isObject(obj: ThereforeTypes | { [schema.type]: string }): obj is ObjectType {
    return obj[schema.type] === 'object'
}

export function isDict(obj: ThereforeTypes | { [schema.type]: string }): obj is DictType {
    return obj[schema.type] === 'dict'
}

export function isArray(obj: ThereforeTypes | { [schema.type]: string }): obj is ArrayType {
    return obj[schema.type] === 'array'
}

export function isTuple(obj: ThereforeTypes | { [schema.type]: string }): obj is TupleType {
    return obj[schema.type] === 'tuple'
}

export function isRef(obj: ThereforeTypes | { [schema.type]: string }): obj is RefType {
    return obj[schema.type] === '$ref'
}

export function isUnion(obj: ThereforeTypes | { [schema.type]: string }): obj is UnionType {
    return obj[schema.type] === 'union'
}

export function isIntersection(obj: ThereforeTypes | { [schema.type]: string }): obj is IntersectionType {
    return obj[schema.type] === 'intersection'
}

export function isExportable(
    obj: unknown | ThereforeTypes
): obj is TupleType | DictType | ObjectType | EnumType | UnionType | IntersectionType {
    return isShorthand(obj) //&&(isTuple(obj) || isDict(obj) || isObject(obj) || isEnum(obj) || isRef(obj) || isUnion(obj) || isIntersection(obj))
}

export function getSchemaFilename(file: string, extension: string): string {
    return `${file.replace('.ts', '')}${extension}`
}

export function requireReference(
    definitions: Record<string, FileDefinition>,
    current: FileDefinition,
    ref: TypescriptReference
): { file: string; type: string; uuid: string; reference: string } {
    const found = Object.values(definitions)
        .map((d) =>
            d.symbols.map((s) => ({
                uuid: s.tsDefinition.uuid,
                interfaceName: s.tsDefinition.interfaceName,
                referenceName: s.tsDefinition.referenceName,
                file: d.file,
            }))
        )
        .flat()
        .find((s) => s.uuid === ref.uuid)

    if (!found) {
        // this is a locally defined variable
        const tsDefinition = toTypescriptDefinition(ref.name, ref.reference, false)
        current.symbols.push({
            name: ref.name,
            tsDefinition: tsDefinition,
        })

        for (const ref of tsDefinition.references) {
            // recurse down to referencing references
            requireReference(definitions, current, ref)
        }

        //return undefined
        //throw Error(`reference ${ref.uuid} was not found`)
        return {
            file: current.file,
            type: tsDefinition.interfaceName,
            reference: tsDefinition.referenceName,
            uuid: ref.uuid,
        }
    }
    return {
        file: found.file,
        type: found.interfaceName,
        reference: found.referenceName,
        uuid: ref.uuid,
    }
}
// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
const requireFunc = typeof __webpack_require__ === 'function' ? __non_webpack_require__ : require
export function scanModule(entry: string, definitions: { [k: string]: FileDefinition }, fullPath: string): void {
    const definition: FileDefinition = {
        file: fullPath,
        jsonFiles: [],
        symbols: [],
        dependencies: {},
    }
    const relative = path.relative(__dirname, entry).replace(/\\/g, '/').replace('.ts', '')
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call
    const module = requireFunc(relative.startsWith('.') ? relative : `./${relative}`)
    for (const [name, symbol] of Object.entries(module)) {
        if (isExportable(symbol)) {
            const schemaName = decamelize(name, '-')
            const schemaFile = `./schemas/${schemaName}.schema.json`
            const schema = toJsonSchema(symbol)

            const tsDefinition = toTypescriptDefinition(name, symbol)
            definition.symbols.push({
                name, // root: path.dirname(entry),
                //schema,
                tsDefinition: tsDefinition,
                schemaFile,
            })

            if (tsDefinition.meta !== undefined) {
                const file = path.join(path.dirname(entry), schemaFile)
                definition.jsonFiles.push({ file, schema: JSON.stringify(schema.schema, null, 2) })
            }

            console.log(` - found ${tsDefinition.interfaceName}`)
        }
        if (definition.symbols.length) {
            definitions[fullPath] = definition
        }
    }
}

export function scanFiles(files: string[], basePath: string): { [k: string]: FileDefinition } {
    const definitions: {
        [k: string]: FileDefinition
    } = {}
    for (const entry of files) {
        const fullPath = path.resolve(basePath, entry)
        let isGenerated = false
        const strippedPath = fullPath.substr(0, fullPath.lastIndexOf('.'))
        for (const ext of ['.ts', '.js']) {
            const p = `${strippedPath}${ext}`
            isGenerated ||= fs.existsSync(p) && fs.readFileSync(p).includes(generatedBy)
        }
        if (isGenerated) {
            console.debug(`scanning ${entry}`)
            console.debug(` * skipping generated schema`)
            continue
        }

        console.log(`scanning ${entry}`)

        try {
            scanModule(entry, definitions, fullPath)
        } catch (e) {
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
            console.debug(e.message, e.stack)
            throw e
        }
    }
    return definitions
}

// eslint-disable-next-line @typescript-eslint/require-await
export async function compileSchemas(
    entries: string[],
    extension: string,
    cwd: string
): Promise<{ file: string; template: string; data: Record<string, string>; type: 'typescript' | 'jsonschema' }[]> {
    const schemaFiles: { file: string; template: string; data: Record<string, string>; type: 'typescript' | 'jsonschema' }[] = []
    const definitions = scanFiles(entries, cwd)

    // const localReferences: Record<string, string> = Object.fromEntries(
    //     Object.values(definitions)
    //         .map((d) => d.symbols.map((s) => s.tsDefinition.references.map((r) => [r.uuid, r.referenceName])))
    //         .flat(2)
    // ) as Record<string, string>
    const localReferences = {}

    for (const [file, definition] of Object.entries(definitions)) {
        const required = definition.symbols
            .map((i) => i.tsDefinition.references.map((r) => requireReference(definitions, definition, r)))
            .flat()
        //.filter(isDefined)

        for (const r of required) {
            definition.dependencies[r.file] ??= []
            definition.dependencies[r.file]?.push(r.type)
        }

        const schemaFileName = getSchemaFilename(file, extension)

        const relativeDependencies = Object.entries(definition.dependencies)
            .map(([f, deps]) => ({
                file: getSchemaFilename(f, extension),
                deps: `{ ${[...new Set(deps)].sort().join(', ')} }`,
            }))
            .filter(({ file }) => file !== schemaFileName)
            .map(({ file, deps }) => {
                const otherPath = path.relative(path.dirname(schemaFileName), file).replace('.ts', '').replace(/\\/g, '/')
                return `import ${deps} from '${otherPath.startsWith('.') ? otherPath : `./${otherPath}`}'`
            })

        const references = Object.fromEntries(required.map((r) => [r.uuid, r.reference] as [string, string]))
        schemaFiles.push(
            {
                file: schemaFileName,
                type: 'typescript',
                template: ([] as string[])
                    .concat(
                        ['/**', ` * ${generatedBy}@v${version as string}`, ' * Do not manually touch this', ' */\n'],
                        ['/* eslint-disable */'],
                        definition.symbols
                            .filter((i) => i.tsDefinition?.meta !== undefined && i.schemaFile)
                            .map((i) => `import ${i.name}Schema from '${i.schemaFile!}'`)
                            .sort(),
                        relativeDependencies.length ? ['', ...relativeDependencies] : [],
                        ['', `import AjvValidator from 'ajv'`, ''],
                        Object.values(definition.symbols)
                            .map((i) => [
                                i.tsDefinition.declaration,
                                i.tsDefinition.meta !== undefined
                                    ? renderTemplate(i.tsDefinition.meta, {
                                          schema: `${i.name}Schema`,
                                      })
                                    : undefined,
                            ])
                            .flat(2)
                            .filter(isDefined)
                    )
                    .join('\n'),
                data: { ...localReferences, ...references },
            },
            ...definition.jsonFiles.map((j) => ({
                file: j.file,
                type: 'jsonschema' as const,
                template: j.schema,
                data: { ...localReferences, ...references },
            }))
        )
    }

    return schemaFiles
}

export async function execute({
    globs,
    excludes,
    format,
    extension,
}: {
    globs: string[]
    excludes: string[]
    format: boolean
    extension: string
}): Promise<void> {
    console.log(globs)
    const entries = await fastGlob([...globs, ...excludes.map((e) => `!${e}`)], { dot: true })

    const schemaFiles = await compileSchemas(entries, extension, process.cwd())

    for (const { file, template, data } of schemaFiles) {
        //fs.writeFileSync(file)
        fs.mkdirSync(path.dirname(file), { recursive: true })
        const references = {}
        console.log(file, references)
        fs.writeFileSync(file, renderTemplate(template, data))

        // for (const { file, schema } of definition.jsonFiles) {
        //     fs.mkdirSync(path.dirname(file), { recursive: true })
        //     console.log(file, references)
        //     fs.writeFileSync(file, template(schema, references))
        // }
    }

    if (schemaFiles.length > 0 && format) {
        execa.commandSync(`yarn prettier --write ${schemaFiles.map((f) => path.relative(process.cwd(), f.file)).join(' ')}`, {
            stdio: 'inherit',
        })
    }
}
