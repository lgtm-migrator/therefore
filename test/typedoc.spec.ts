jest.mock('uuid')

import { mockUuid } from './util'

import { compileSchemas } from '~/cli'

import { v4 as uuid } from 'uuid'

describe('schema', () => {
    beforeEach(() => (uuid as jest.Mock).mockImplementation(mockUuid()))
    test('typedoc', async () => {
        expect(await compileSchemas(['examples/typedoc/typedoc.schema.ts'], '.type.ts', process.cwd())).toMatchSnapshot()
    })
})
