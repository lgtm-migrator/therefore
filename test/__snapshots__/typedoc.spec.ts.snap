// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema typedoc 1`] = `
Array [
  Object {
    "data": Object {
      "00010-000": "SrcLocal",
      "00013-000": "ThemeLocal",
      "0003-000": "PluginLocal",
      "0006-000": "ReadmeLocal",
    },
    "file": "examples/typedoc/typedoc.schema.type.ts",
    "prettify": true,
    "relativeSource": "./typedoc.schema",
    "template": "/**
 * Generated by @zefiros-software/therefore@v0.0.1
 * Do not manually touch this
 */

/* eslint-disable */

import type { ValidateFunction } from 'ajv'


import AjvValidator from 'ajv'

/**
 * JSON Schema for typedoc.json
 */
export interface Typedoc {
    /**
     * Should TypeDoc disable the testing and cleaning of the output directory?
     * 
     * @default false
     */
    disableOutputCheck: boolean
    /**
     * Specifies the fully qualified name of the root symbol.
     * 
     * @default './'
     */
    entryPoint: string
    /**
     * Exclude files by the given pattern when a path is provided as source. Supports minimatch patterns.
     */
    exclude: (string)[]
    /**
     * Prevent externally resolved TypeScript files from being documented.
     * 
     * @default false
     */
    excludeExternals: boolean
    /**
     * Prevent symbols that are not exported from being documented.
     * 
     * @default false
     */
    excludeNotExported: boolean
    /**
     * Ignores private variables and methods
     * 
     * @default false
     */
    excludePrivate: boolean
    /**
     * Ignores protected variables and methods
     * 
     * @default false
     */
    excludeProtected: boolean
    /**
     * Define a pattern for files that should be considered being external.
     */
    externalPattern: (string)[]
    /**
     * Set the Google Analytics tracking ID and activate tracking code.
     */
    gaID: string
    /**
     * Set the site name for Google Analytics.
     * 
     * @default 'auto'
     */
    gaSite: string
    /**
     * Use specified revision or branch instead of the last revision for linking to GitHub source files.
     */
    gitRevision: string
    /**
     * Do not print the TypeDoc link at the end of the page.
     * 
     * @default false
     */
    hideGenerator: boolean
    /**
     * Generates documentation, even if the project does not TypeScript compile.
     * 
     * @default false
     */
    ignoreCompilerErrors: boolean
    /**
     * Turn on parsing of .d.ts declaration files.
     * 
     * @default false
     */
    includeDeclarations: boolean
    /**
     * Specifies the location to look for included documents (use [[include:FILENAME]] in comments).
     */
    includes: string
    /**
     * The sources files from which to build documentation.
     */
    inputFiles: (string)[]
    /**
     * Specifies the location to output a JSON file containing all of the reflection data.
     */
    json: string
    /**
     * Emits a list of broken symbol [[navigation]] links after documentation generation
     * 
     * @default false
     */
    listInvalidSymbolLinks: boolean
    /**
     * Specify the logger that should be used.
     * 
     * @default 'console'
     */
    logger: 'console' | 'none'
    /**
     * Specifies the location with media files that should be copied to the output directory.
     */
    media: string
    /**
     * Specifies the output mode the project is used to be compiled with.
     * 
     * @default 'modules'
     */
    mode: 'file' | 'modules'
    /**
     * Set the name of the project that will be used in the header of the template.
     */
    name: string
    /**
     * Specifies the location the documentation should be written to.
     */
    out: string
    plugin: {{0003-000}}
    /**
     * Add the package version according to package.json to the projects name.
     * 
     * @default false
     */
    includeVersion: boolean
    /**
     * Specify tags that should be removed from doc comments when parsing.
     * 
     * @default []
     */
    excludeTags: (string)[]
    readme: {{0006-000}}
    src: {{00010-000}}
    /**
     * Remove reflections annotated with @internal
     * 
     * @default false
     */
    stripInternal: boolean
    theme: {{00013-000}}
    /**
     * Specifies the top level table of contents.
     */
    toc: (string)[]
    /**
     * Specify a typescript config file that should be loaded. If not specified TypeDoc will look for 'tsconfig.json' in the current directory.
     * 
     * @default './tsconfig.json'
     */
    tsconfig: string
}

export const Typedoc = {
    validate: require('./schemas/typedoc.schema.js') as ValidateFunction<Typedoc>,
    get schema() { return Typedoc.validate.schema},
    source: \`\${__dirname}typedoc.schema\`,
    sourceSymbol: 'typedoc',
    is: (o: unknown): o is Typedoc => Typedoc.validate(o) === true,
    assert: (o: unknown): asserts o is Typedoc => {
        if (!Typedoc.validate(o)) {
            throw new AjvValidator.ValidationError(Typedoc.validate.errors ?? [])
        }
    },
} as const

/**
 * Specify the npm plugins that should be loaded. Omit to load all installed plugins.
 */
type PluginLocal = (string)[]

/**
 * Path to the readme file that should be displayed on the index page. Pass none to disable the index page and start the documentation on the globals page.
 */
type ReadmeLocal = 'none' | string

/**
 * The sources files from which to build documentation.
 * DEPRECATED: Use inputFiles instead.
 */
type SrcLocal = string | (string)[]

/**
 * Specify the path to the theme that should be used.
 */
type ThemeLocal = 'default' | 'minimal' | string
",
    "type": "typescript",
  },
  Object {
    "data": Object {
      "00010-000": "SrcLocal",
      "00013-000": "ThemeLocal",
      "0003-000": "PluginLocal",
      "0006-000": "ReadmeLocal",
    },
    "file": "examples/typedoc/schemas/typedoc.schema.json",
    "prettify": true,
    "relativeSource": "./typedoc.schema",
    "template": "{
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"type\\": \\"object\\",
  \\"title\\": \\"JSON Schema for typedoc.json\\",
  \\"properties\\": {
    \\"disableOutputCheck\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Should TypeDoc disable the testing and cleaning of the output directory?\\",
      \\"default\\": false
    },
    \\"entryPoint\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Specifies the fully qualified name of the root symbol.\\",
      \\"default\\": \\"./\\"
    },
    \\"exclude\\": {
      \\"type\\": \\"array\\",
      \\"description\\": \\"Exclude files by the given pattern when a path is provided as source. Supports minimatch patterns.\\",
      \\"items\\": {
        \\"type\\": \\"string\\"
      }
    },
    \\"excludeExternals\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Prevent externally resolved TypeScript files from being documented.\\",
      \\"default\\": false
    },
    \\"excludeNotExported\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Prevent symbols that are not exported from being documented.\\",
      \\"default\\": false
    },
    \\"excludePrivate\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Ignores private variables and methods\\",
      \\"default\\": false
    },
    \\"excludeProtected\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Ignores protected variables and methods\\",
      \\"default\\": false
    },
    \\"externalPattern\\": {
      \\"type\\": \\"array\\",
      \\"description\\": \\"Define a pattern for files that should be considered being external.\\",
      \\"items\\": {
        \\"type\\": \\"string\\"
      }
    },
    \\"gaID\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Set the Google Analytics tracking ID and activate tracking code.\\"
    },
    \\"gaSite\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Set the site name for Google Analytics.\\",
      \\"default\\": \\"auto\\"
    },
    \\"gitRevision\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Use specified revision or branch instead of the last revision for linking to GitHub source files.\\"
    },
    \\"hideGenerator\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Do not print the TypeDoc link at the end of the page.\\",
      \\"default\\": false
    },
    \\"ignoreCompilerErrors\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Generates documentation, even if the project does not TypeScript compile.\\",
      \\"default\\": false
    },
    \\"includeDeclarations\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Turn on parsing of .d.ts declaration files.\\",
      \\"default\\": false
    },
    \\"includes\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Specifies the location to look for included documents (use [[include:FILENAME]] in comments).\\"
    },
    \\"inputFiles\\": {
      \\"type\\": \\"array\\",
      \\"description\\": \\"The sources files from which to build documentation.\\",
      \\"items\\": {
        \\"type\\": \\"string\\"
      }
    },
    \\"json\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Specifies the location to output a JSON file containing all of the reflection data.\\"
    },
    \\"listInvalidSymbolLinks\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Emits a list of broken symbol [[navigation]] links after documentation generation\\",
      \\"default\\": false
    },
    \\"logger\\": {
      \\"description\\": \\"Specify the logger that should be used.\\",
      \\"default\\": \\"console\\",
      \\"enum\\": [
        \\"console\\",
        \\"none\\"
      ]
    },
    \\"media\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Specifies the location with media files that should be copied to the output directory.\\"
    },
    \\"mode\\": {
      \\"description\\": \\"Specifies the output mode the project is used to be compiled with.\\",
      \\"default\\": \\"modules\\",
      \\"enum\\": [
        \\"file\\",
        \\"modules\\"
      ]
    },
    \\"name\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Set the name of the project that will be used in the header of the template.\\"
    },
    \\"out\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Specifies the location the documentation should be written to.\\"
    },
    \\"plugin\\": {
      \\"$ref\\": \\"#/definitions/{{0003-000}}\\"
    },
    \\"includeVersion\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Add the package version according to package.json to the projects name.\\",
      \\"default\\": false
    },
    \\"excludeTags\\": {
      \\"type\\": \\"array\\",
      \\"description\\": \\"Specify tags that should be removed from doc comments when parsing.\\",
      \\"default\\": [],
      \\"items\\": {
        \\"type\\": \\"string\\"
      }
    },
    \\"readme\\": {
      \\"$ref\\": \\"#/definitions/{{0006-000}}\\"
    },
    \\"src\\": {
      \\"$ref\\": \\"#/definitions/{{00010-000}}\\"
    },
    \\"stripInternal\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Remove reflections annotated with @internal\\",
      \\"default\\": false
    },
    \\"theme\\": {
      \\"$ref\\": \\"#/definitions/{{00013-000}}\\"
    },
    \\"toc\\": {
      \\"type\\": \\"array\\",
      \\"description\\": \\"Specifies the top level table of contents.\\",
      \\"items\\": {
        \\"type\\": \\"string\\"
      }
    },
    \\"tsconfig\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Specify a typescript config file that should be loaded. If not specified TypeDoc will look for 'tsconfig.json' in the current directory.\\",
      \\"default\\": \\"./tsconfig.json\\"
    }
  },
  \\"required\\": [
    \\"disableOutputCheck\\",
    \\"entryPoint\\",
    \\"exclude\\",
    \\"excludeExternals\\",
    \\"excludeNotExported\\",
    \\"excludePrivate\\",
    \\"excludeProtected\\",
    \\"externalPattern\\",
    \\"gaID\\",
    \\"gaSite\\",
    \\"gitRevision\\",
    \\"hideGenerator\\",
    \\"ignoreCompilerErrors\\",
    \\"includeDeclarations\\",
    \\"includes\\",
    \\"inputFiles\\",
    \\"json\\",
    \\"listInvalidSymbolLinks\\",
    \\"logger\\",
    \\"media\\",
    \\"mode\\",
    \\"name\\",
    \\"out\\",
    \\"plugin\\",
    \\"includeVersion\\",
    \\"excludeTags\\",
    \\"readme\\",
    \\"src\\",
    \\"stripInternal\\",
    \\"theme\\",
    \\"toc\\",
    \\"tsconfig\\"
  ],
  \\"additionalProperties\\": false,
  \\"definitions\\": {
    \\"{{0003-000}}\\": {
      \\"title\\": \\"plugin\\",
      \\"description\\": \\"Specify the npm plugins that should be loaded. Omit to load all installed plugins.\\",
      \\"oneOf\\": [
        {
          \\"type\\": \\"array\\",
          \\"default\\": [
            \\"none\\"
          ],
          \\"items\\": {
            \\"type\\": \\"string\\"
          }
        }
      ]
    },
    \\"{{0006-000}}\\": {
      \\"title\\": \\"readme\\",
      \\"description\\": \\"Path to the readme file that should be displayed on the index page. Pass none to disable the index page and start the documentation on the globals page.\\",
      \\"oneOf\\": [
        {
          \\"const\\": \\"none\\"
        },
        {
          \\"type\\": \\"string\\"
        }
      ]
    },
    \\"{{00010-000}}\\": {
      \\"title\\": \\"src\\",
      \\"description\\": \\"The sources files from which to build documentation.\\\\nDEPRECATED: Use inputFiles instead.\\",
      \\"oneOf\\": [
        {
          \\"type\\": \\"string\\"
        },
        {
          \\"type\\": \\"array\\",
          \\"items\\": {
            \\"type\\": \\"string\\"
          }
        }
      ]
    },
    \\"{{00013-000}}\\": {
      \\"title\\": \\"theme\\",
      \\"description\\": \\"Specify the path to the theme that should be used.\\",
      \\"oneOf\\": [
        {
          \\"enum\\": [
            \\"default\\",
            \\"minimal\\"
          ]
        },
        {
          \\"type\\": \\"string\\"
        }
      ]
    }
  }
}",
    "type": "jsonschema",
  },
]
`;
